### PR TITLE
- work on exceptions

### PR DESCRIPTION
**Pull request by aschnell**

### CODE REVIEW

File: storage/Devices/BlkDevice.cc [hunk 1]
```diff
@@ -48,7 +48,10 @@ namespace storage
 		return blk_device;
 	}
 
-	throw DeviceNotFound(sformat("device not found, name:%s", name.c_str()));
+	ST_THROW(DeviceNotFound(sformat("device not found, name:%s", name.c_str())));
```
Comment (shundhammer) [inline storage/Devices/BlkDevice.cc]: a blank would make it much better readable:  name: %s

File: storage/Devices/Disk.cc [hunk 2]
```diff
@@ -133,7 +133,7 @@ namespace storage
     {
 	Disk* disk = dynamic_cast<Disk*>(device);
 	if (!disk)
-	    throw DeviceHasWrongType("device is not a Disk");
+	    ST_THROW(DeviceHasWrongType("device is not a Disk"));
```
Comment (shundhammer) [inline storage/Devices/Disk.cc]: How about a dedicated check macro in Exception.h for this situation that seems to appear a lot:

#define ST_CHECK_TYPE(VAR, NAME)  if (!(VAR)) ST_THROW(DeviceHasWrongType("device is not a " #NAME))

File: storage/Devices/Disk.cc [hunk 3]
```diff
@@ -144,7 +144,7 @@ namespace storage
     {
 	const Disk* disk = dynamic_cast<const Disk*>(device);
 	if (!disk)
-	    throw DeviceHasWrongType("device is not a Disk");
+	    ST_THROW(DeviceHasWrongType("device is not a Disk"));
```
Comment (shundhammer) [inline storage/Devices/Disk.cc]: Same as above: ST_CHECK_TYPE

File: storage/Devices/Filesystem.cc [hunk 4]
```diff
@@ -178,7 +178,7 @@ namespace storage
     {
 	Filesystem* filesystem = dynamic_cast<Filesystem*>(device);
 	if (!filesystem)
-	    throw DeviceHasWrongType("device is not a Filesystem");
+	    ST_THROW(DeviceHasWrongType("device is not a Filesystem"));
```
Comment (shundhammer) [inline storage/Devices/Filesystem.cc]: ST_CHECK_TYPE again

File: storage/Devices/Filesystem.cc [hunk 5]
```diff
@@ -189,7 +189,7 @@ namespace storage
     {
 	const Filesystem* filesystem = dynamic_cast<const Filesystem*>(device);
 	if (!filesystem)
-	    throw DeviceHasWrongType("device is not a Filesystem");
+	    ST_THROW(DeviceHasWrongType("device is not a Filesystem"));
```
Comment (shundhammer) [inline storage/Devices/Filesystem.cc]: ST_CHECK_TYPE again

File: storage/Devices/PartitionTable.cc [hunk 6]
```diff
@@ -86,7 +86,7 @@ namespace storage
     {
 	PartitionTable* partition_table = dynamic_cast<PartitionTable*>(device);
 	if (!partition_table)
-	    throw DeviceHasWrongType("device is not a PartitionTable");
+	    ST_THROW(DeviceHasWrongType("device is not a PartitionTable"));
```
Comment (shundhammer) [inline storage/Devices/PartitionTable.cc]: ST_CHECK_TYPE again

File: storage/Devices/PartitionTable.cc [hunk 7]
```diff
@@ -97,7 +97,7 @@ namespace storage
     {
 	const PartitionTable* partition_table = dynamic_cast<const PartitionTable*>(device);
 	if (!partition_table)
-	    throw DeviceHasWrongType("device is not a PartitionTable");
+	    ST_THROW(DeviceHasWrongType("device is not a PartitionTable"));
```
Comment (shundhammer) [inline storage/Devices/PartitionTable.cc]: ST_CHECK_TYPE again

File: storage/Devices/DiskImpl.cc [hunk 8]
```diff
@@ -150,28 +150,14 @@ namespace storage
...
     {
-	if (get_device()->num_children() != 1)
-	    throw runtime_error("disk has no children");
-
-	const Devicegraph* devicegraph = get_devicegraph();
-
-	Device* child = devicegraph->get_impl().graph[devicegraph->get_impl().child(get_vertex())].get();
-
-	return to_partition_table(child);
+	return get_single_child_of_type<PartitionTable>();
```
Comment (shundhammer) [inline storage/Devices/DiskImpl.cc]: Big improvement over the original code!

**Comment by shundhammer** [comment]
Other than the commented parts (which are all on the nitpicking level), LGTM.

File: storage/Devices/Disk.cc (hunk 2 above)
Comment (aschnell) [inline storage/Devices/Disk.cc]: Yes, that is used often. I also realized. But sometimes the exception is of type DeviceHasWrongType and sometimes of HolderHasWrongType. My idea it to use a templated function that replaces the complete function body.

File: storage/Devices/BlkDevice.cc (hunk 1 above)
Comment (aschnell) [inline storage/Devices/BlkDevice.cc]: OK, I'm used to that style from the old libstorage. There "key1:value1 key2:value2" is used. I'm fine with using "key1: value1, key2: value2" in the future.

**Pull request merged**